[Codegen][GPU] Add range information to GPU dispatch IDs

krzysz00 opened the pull request:
**NOTE:** Don't land this until upstream PR  https://github.com/llvm/llvm-project/pull/95166 lands so I can use `gpu.thread_id x upper_bound X`.

First, this patch implements InferIntRangeInterface for hal.interface.workgroup.{size,id,count} using a local upper_bound attribute.

Then, it adds a -iree-codegen-gpu-propagate-dispatch-size-bounds pass that adds these upper_bounds identifiers to the interface.workgroup operations and to gpu.thread_id based on static information available late in the codegen pipeline.

Then, it uses -optimize-int-arithmetic to optimize indexing after -lower-affine, getting rid of a bunch of "if the input's negative" logic that isn't actually needed in many of our kernels.

It also ensures that these upper_bound values propagate to LLVM (at least on ROCDL - NVVM doesn't have this hooked up to the intrinsics).

kuhar reviewed (changes requested):
Could we separate general improvements to the rocm pipeline from new functionality (range information)? This will make it easier to review and later bisect.

krzysz00 committed:
[Codegen][GPU] Add range information to GPU dispatch IDs

First, this patch implements InferIntRangeInterface for hal.interface.workgroup.{size,id,count} using a local upper_bound attribute.

Then, it adds a -iree-codegen-gpu-propagate-dispatch-size-bounds pass that adds these upper_bounds identifiers to the interface.workgroup operations and to gpu.thread_id based on static information available late in the codegen pipeline.

Then, it uses -optimize-int-arithmetic to optimize indexing after
-lower-affine, getting rid of a bunch of "if the input's negative"
logic that isn't actually needed in many of our kernels.

It also ensures that these upper_bonud values propagate to LLVM (at least on ROCDL - NVVM doesn't have this hooked up to the intrinsics).

Signed-off-by: Krzysztof Drewniak <Krzysztof.Drewniak@amd.com>

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+                        ArrayRef<int64_t> workgroupCounts) {
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    if (auto tidOp = dyn_cast<gpu::ThreadIdOp>(op)) {
```
Use type switch

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+namespace {
+
+static void applyBounds(FunctionOpInterface funcOp,
+                        ArrayRef<int64_t> workitemCounts,
```
I don't think we use the name 'workitems' in codegen. Is this the same as workgroup size?

krzysz00 replied:
Yeah, this is workgroup size, but I was hazy about `workgroupSizes` and `workgroupCounts` in the same context. Not a big change if you don't think it'll confuse anyone

kuhar replied:
I'd prefer to stay consistent with the rest of the codebase

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+      GPUPropagateDispatchSizeBoundsPassBase;
+  void runOnOperation() override {
+    FunctionOpInterface funcOp = getOperation();
+    SmallVector<int64_t> workitemCounts(3, maxWorkgroupSize);
```
Same here

kuhar replied:
nit: Since the size is statically known, we can make it an array. Also below.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+    SmallVector<int64_t> workgroupCounts = {maxWorkgroupsX, maxWorkgroupsYZ,
+                                            maxWorkgroupsYZ};
+
+    std::optional<SmallVector<int64_t>> staticWorkitemCounts =
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+      if (staticWorkitemCounts->size() > 3) {
+        // How to flag this but not fail the pipeline?
+        LLVM_DEBUG(llvm::dbgs()
+                   << "more than 3 workgroup size dimensions in GPU code\n");
```
`funcOp.emitWarning`

krzysz00 replied:
Ah, that's the right style, thanks, will fix

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+      return;
+    }
+    for (auto [known, upperBound] :
+         llvm::zip(staticWorkgroupCounts, workgroupCounts))
```
zip_equals

krzysz00 replied:
I'm not sure the static workgroup bound getter is guaranteed to return exactly three results? Unless I missed a comment

kuhar replied:
Oh I see what you mean. Would be nice to capture this in a test.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/Passes.td:
```diff
@@ -111,6 +111,22 @@ def GPUPromoteMatmulOperandsPass :
+          "Maximum number of workgroups in the y and z dimension permitted by the hardware">,
+    Option<"maxWorkgroupSize", "max-workgroup-size", "unsigned",
+          /*default=*/"1024",
+          "Maximum workgroup size when none is specified">
```
BTW, we already have some target device info like this in `hal.executable.variant` target info:
```
    hal.executable.variant public @rocm_hsaco_fb target(<"rocm", "rocm-hsaco-fb", {iree.gpu.target = #iree_gpu.target<arch = "gfx942", features = "", wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8, storage =  b64|b32|b16|b8, subgroup =  shuffle|arithmetic, dot =  dp4xi8toi32, mma = [<MFMA_F16_16x16x16_F32>, <MFMA_F16_32x32x8_F32>], subgroup_size_choices = [64], max_workgroup_sizes = [1024, 1024, 1024], max_thread_count_per_workgroup = 1024, max_workgroup_memory_bytes = 65536>>, ukernels = "none", waves_per_eu = 2 : i64}>) {
```

krzysz00 replied:
Good point, I could just get it from there

qedawkins replied:
+1 all of these values should be queried from the target information annotating the parent function rather than being pass options.

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,87 @@
+    for (auto [known, upperBound] :
+         llvm::zip(staticWorkgroupCounts, workgroupCounts))
+      if (known != ShapedType::kDynamic)
+        upperBound = known;
```
Style nit: IREE style is to always add braces for nested statements, unlike LLVM.

qedawkins reviewed (commented):
I'm happy to approve this, but @stellaraccident has recently been adding some util ops for annotating ranges like this, e.g. https://github.com/iree-org/iree/commit/462ecb691a523a1b329cd14707fdcbed1d85f116

Do we want to avoid adding to the HAL ops themselves if things like `util.assume.range` and `util.assume.divisible` are available? If there isn't a strong sentiment against adding to the workgroup ID/size ops then I have no strong feeling one way or the other.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,95 @@
+                        ArrayRef<int32_t> workgroupCounts) {
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    TypeSwitch<Operation *, void>(op)
```
nit: `void` is the default
```suggestion
    TypeSwitch<Operation *>(op)
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,95 @@
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    TypeSwitch<Operation *, void>(op)
+        .Case<gpu::ThreadIdOp>([&](auto tidOp) {
```
These don't have to use `auto` since the case covers a single op type. Same below.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,95 @@
+          wgCountOp.setUpperBoundAttr(b.getIndexAttr(
+              workgroupCounts[wgCountOp.getDimension().getZExtValue()]));
+        })
+        .Default([](Operation *op) { std::ignore = op; });
```
```suggestion
        .Default([](Operation *) {});
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,95 @@
+    }
+    SmallVector<int64_t> staticWorkgroupCounts = getStaticNumWorkgroups(funcOp);
+    if (staticWorkgroupCounts.size() > 3) {
+      funcOp.emitWarning("more than 3 workgroup count dimensions");
```
Shouldn't this be an assertion?

krzysz00 replied:
Not sure. I don't know if really multi-dimensional dispatches are invalid IR

kuhar replied:
Can we assert here then if we are not sure?

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_propagate_dispatch_size_bounds.mlir:
```diff
@@ -0,0 +1,121 @@
+// RUN: iree-opt %s --split-input-file --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(builtin.module(func.func(iree-codegen-gpu-propagate-dispatch-size-bounds)))))" | FileCheck %s
```
Could you break this line to make it a little bit more readable?

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp:
```diff
@@ -495,7 +495,10 @@ struct HALInterfaceWorkgroupOpsConverter final
-    rewriter.replaceOpWithNewOp<NewOpTy>(op, op.getType(), dimAttr[index]);
+    NewOpTy newOp =
+        rewriter.replaceOpWithNewOp<NewOpTy>(op, op.getType(), dimAttr[index]);
+    if (auto bound = op.getUpperBoundAttr())
```
nit: The type is not clear to me, could you use the full attribute type here?

kuhar reviewed (commented):
Could you click the 'resolve conversation' button on the past comments that have been addressed?

krzysz00 committed:
Review feedback

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,108 @@
+              exportOp->getWorkgroupSize()) {
+        staticWorkgroupSize =
+            llvm::map_to_vector(exportWorkgroupSize->getAsRange<IntegerAttr>(),
+                                [](auto a) { return a.getInt(); });
```
```suggestion
                                [](IntegerAttr a) { return a.getInt(); });
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,108 @@
+    : impl::GPUPropagateDispatchSizeBoundsPassBase<
+          GPUPropagateDispatchSizeBoundsPass> {
+  using GPUPropagateDispatchSizeBoundsPassBase::
+      GPUPropagateDispatchSizeBoundsPassBase;
```
I think this should also work
```suggestion
  using Base::Base;
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_propagate_dispatch_size_bounds.mlir:
```diff
@@ -0,0 +1,121 @@
+#executable_target = #hal.executable.target<"rocm", "rocm-hsaco-fb", {iree.gpu.target = #iree_gpu.target<arch = "gfx1100", features = "",
+  wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8,
+    storage =  b64|b32|b16|b8,
+    subgroup =  shuffle|arithmetic,
```
Could you prune these attributes a bit? For example, I don't think the test needs to know about all these types

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_propagate_dispatch_size_bounds.mlir:
```diff
@@ -0,0 +1,121 @@
+  wgp = <compute =  fp64|fp32|fp16|int64|int32|int16|int8,
+    storage =  b64|b32|b16|b8,
+    subgroup =  shuffle|arithmetic,
+    dot =  dp4xi8toi32, mma = [<WMMA_F32_16x16x16_F16>, <WMMA_F16_16x16x16_F16>, <WMMA_I32_16x16x16_I8>],
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-propagate-dispatch-size-bounds"
```
This appears unused?

kuhar reviewed (commented):
Looks good overall, just a few suggestions

krzysz00 committed:
Review comments

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_propagate_dispatch_size_bounds.mlir:
```diff
@@ -0,0 +1,122 @@
+  {iree.gpu.target = #iree_gpu.target<arch = "gfx1100", features = "",
+  wgp = <compute =  fp32,
+    storage =  b32,
+    subgroup =  shuffle|arithmetic,
```
```suggestion
    subgroup = arithmetic,
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_propagate_dispatch_size_bounds.mlir:
```diff
@@ -0,0 +1,122 @@
+#executable_target = #hal.executable.target<"rocm", "rocm-hsaco-fb", {iree.gpu.target = #iree_gpu.target<arch = "gfx1100", features = "",
+  wgp = <compute =  fp32,
+    storage =  b32,
+    subgroup =  shuffle|arithmetic,
```
```suggestion
    subgroup =  arithmetic,
```

kuhar reviewed (approved):
LGTM % one nit

krzysz00 committed:
Simplify tests per last nit

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>
